rtio: Transactional request support

teburd opened the pull request:
Adds the ability to mark submissions as being transactional which has the unique semantics of allowing multiple receive/send type operations to be viewed by the submitted to iodev as a single unit of work to do.

This is needed for SPI where a sequence of reads/writes with chip select are part of a transaction.

Commit:
rtio: Fix consume semaphore usage

When the consume semaphore is enabled always give and take from the
semaphore. It's expected that rtio_cqe_consume and rtio_cqe_consume_block
will be used exclusively rather than directly poking at the SPSC queues.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Adds rtio_iodev_cancel_all

Ability to cancel all pending requests in the queue turns out to be
useful and shareable.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Add transactional submissions

Transactional submissions treat a sequence of sqes as a single atomic
submission given to a single iodev and expect as a reply a single
completion.

This is useful for scatter gather like APIs that exist in Zephyr already
for I2C and SPI.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Clear flags in prep helpers

Tests were failing when the sqe was allocated on the stack for some
architectures. Initialization of variables on the stack is not guaranteed
to be zeroed like static data. This caused undefined behavior.

Secondly if the sqe is recycled and had a flag set, there would
unexpected behavior as well.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
tests: rtio: Fix potential race in iodev test

Race was possible though very unlikely between the atomic cas
and queue push/pop operations. The outcome of the race had it shown up
would have been a submission not worked on due to the timer never being
started. A small critical section fixes this and clarifies where the single
conumer part of the mpsc queue comes in despite there being multiple
contexts which may enter that section.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
tests: rtio: Split test suites up into files

The test suites have grown to cover different units really and having
them in one file was becoming a bit much to scroll around in.
Coincidentally found a accidental reuse of a define between the spsc and
mpsc tests.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Make rtio_mpsc.h C++ friendly

C++ requires casting void * as the implicit cast isn't enough and
the C++ sample app fails to build without duplicating type information
here. Do that, and wrap it in extern C allowing the C++ to include
rtio_mpsc.h directly or indirectly.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

yperess reviewed (dismissed):
Now that I understand this a bit better I'm happy to approve :)

Commit:
rtio: Properly track last sqe in the queue

The pending_sqe logic to track where in the ring queue the concurrent
executor had left off was slightly flawed. It didn't account for starting
all sqes in the queue and ending back up at the beginning.

Instead track the last SQE in the queue, from which the next one in the
queue will the one to start next.

If we happen to sweep the last known SQE in the queue, reset it to NULL
so the next time prepare is called we start at the beginning of the queue
again.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Includes a stack of bug fixes now so CI hopefully remains happy

yperess reviewed (approved):
Verified locally
